torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,366 @@
+        """
+        Used for implementing aten ops by applying them only to the relevant tensor atributes
+        In this case we only want to call things like to() or view() on the layout tensor
+        """
+        return self.__class__(
+            fn(self.layout_tensor),
+            self.shape,
+            self.dtype,
+        )
+
+    @classmethod
+    def implements(cls, aten_ops_or_torch_fns):  # type: ignore[no-untyped-def]
+        if not hasattr(cls, "_DISPATCH_TABLE"):
+            cls._DISPATCH_TABLE = {}  # type: ignore[attr-defined]
+
+        if not isinstance(aten_ops_or_torch_fns, (list, tuple)):
+            aten_ops_or_torch_fns = [aten_ops_or_torch_fns]
+
+        def decorator(func):  # type: ignore[no-untyped-def]
+            for op in aten_ops_or_torch_fns:
+
+                @functools.wraps(op)
+                def wrapper(f, types, args, kwargs):  # type: ignore[no-untyped-def]
+                    return func(f, types, args, kwargs)
+
+                cls._DISPATCH_TABLE[op] = wrapper  # type: ignore[attr-defined]
+
+            return func
+
+        return decorator
+
+    """There are two entry points that we can modify the behavior of a pytorch op: torch_function and torch_dispatch:
+
+    __torch_function__: will be called whenever a torch level function is called on the Tensor object,
+    for example: torch.nn.functional.linear, tensor.detach, tensor.reshape, tensor.t etc.
+
+    __torch_dispatch__: will be called in the C++ dispatcher, when an aten operator is called on the Tensor object,
+    for example: aten.mm, aten.addmm, aten.detach.default, aten.t.default etc.
+
+    We have some helper functions that can dispatch to the functions registered with MyDTypeTensor.implements,
+    but if the default implementation does not work for your use case, please feel free to customize it
+    """
+
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs=None):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if hasattr(cls, "_DISPATCH_TABLE") and func in cls._DISPATCH_TABLE:
+            return cls._DISPATCH_TABLE[func](func, types, args, kwargs)
+
+        with torch._C.DisableTorchFunctionSubclass():
+            return func(*args, **kwargs)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if hasattr(cls, "_DISPATCH_TABLE") and func in cls._DISPATCH_TABLE:
+            return cls._DISPATCH_TABLE[func](func, types, args, kwargs)
+
+        raise NotImplementedError(
+            f"{cls.__name__} dispatch: attempting to run unimplemented operator/function: {func}"
+        )
+
+
+######################################################
+# LayoutType and Layout Tensor Subclass Registration #
+######################################################
+
+
+class PlainMyDTypeLayout(MyDTypeLayout):
```

Comment:
I'd prefer to keep this class simple-and-modifiable over flexible-and-complicated.

In that vein, can we remove all of the multiple layers of inheritance, and just have a single subclass that inherits from `torch.Tensor` and exposes all of its logic?

Reply:
In a similar vein, I was expecting a more specialized quantization example here. What flavor of quantization is this subclass doing? (what's the hardcoded dtype, is it hardcoded to row-wise quantization, etc).

Reply:
multiple layers of inheritance is actually what broke tensor_subclass + torch.compile for a lot of pytorch versions in the past actually, do we really want to remove it?

Reply:
sure I can update the name to be more specialized

Reply:
yeah I think I would prefer to remove it. IIRC, most of the "bugs" we saw with multiple inheritance were actaully bugs in the sublcass impl, as opposed to bugs in the compiler that we needed to fix to get multiple inheritance working